### ISSUE TITLE
Try Agglomerative clustering with average linkage and cosine distance

### ISSUE THREAD

**Issue by dgarrick** [dgarrick/headliner#2]
http://scikit-learn.sourceforge.net/dev/modules/generated/sklearn.cluster.AgglomerativeClustering.html#sklearn.cluster.AgglomerativeClustering

http://scikit-learn.sourceforge.net/dev/modules/clustering.html#hierarchical-clustering

Right now, the Kmeans clustering is OK, but I'm seeing similar results as: https://www.kaggle.com/c/word2vec-nlp-tutorial/details/part-3-more-fun-with-word-vectors

Some clusters are closely related, while others seem completely random. It could be a result of our dataset, but I think the fact that Kmeans uses euclidian distance might be causing problems, since that's not how word2vec itself relates vectors.

Agglomerative clustering will produce some large clusters and many smaller clusters. Thinking about news stories, this seems to fit our data set: popular topics will have many different articles, while less popular topics will only have a few.

Since it's hierarchical by nature, this would also allow us to rank clusters by size: the bigger a cluster, the more stories around a certain topic. So, we could rank clusters by how popular their topics are.

scitkit's agglomerative clustering also supports cosine distance metric, whereas kmeans only support euclidian distance. We want cosine.

However, agglomerative clustering is slower than kmeans. We might want to move towards a design where we can asynchronously run clustering whenever a certain amount of new data has been accumulated.

If that doesn't work out, we should probably write our own kmeans using word2vec's cosine distance function. However, we'll miss out on scikit's cython optimizations unless we do it the hard way.

Aside from that, our best bet might be to fork scikit-learn and add the ability for kmeans clustering with cosine distance instead of euclidian: https://github.com/scikit-learn/scikit-learn/issues/848
